src/System.Net.Sockets/src/System/Net/Sockets/UnixDomainSocketEndPoint.cs

```diff
@@ -0,0 +1,140 @@
+
+        public UnixDomainSocketEndPoint(string path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
+
+            // Pathname socket addresses should be null-terminated.
+            // Linux abstract socket addresses start with a zero byte, they must not be null-terminated.
+            bool isAbstract = IsAbstract(path);
+            int bufferLength = s_pathEncoding.GetByteCount(path);
+            if (!isAbstract)
+            {
+                // for null terminator
+                bufferLength++;
+            }
+
+            if (path.Length == 0 || bufferLength > s_nativePathLength)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(path), path,
+                    SR.Format(SR.ArgumentOutOfRange_PathLengthInvalid, path, s_nativePathLength));
+            }
+
+            _path = path;
+            _encodedPath = new byte[bufferLength];
+            int bytesEncoded = s_pathEncoding.GetBytes(path, 0, path.Length, _encodedPath, 0);
+            Debug.Assert(bufferLength - (isAbstract ? 0 : 1) == bytesEncoded);
+
+            if (!s_udsSupported.Value)
+            {
+                throw new PlatformNotSupportedException();
+            }
+        }
+
+        internal UnixDomainSocketEndPoint(SocketAddress socketAddress)
+        {
+            if (socketAddress == null)
+            {
+                throw new ArgumentNullException(nameof(socketAddress));
+            }
+
+            if (socketAddress.Family != EndPointAddressFamily ||
+                socketAddress.Size > s_nativeAddressSize)
+            {
+                throw new ArgumentOutOfRangeException(nameof(socketAddress));
+            }
+
+            if (socketAddress.Size > s_nativePathOffset)
+            {
+                _encodedPath = new byte[socketAddress.Size - s_nativePathOffset];
+                for (int i = 0; i < _encodedPath.Length; i++)
+                {
+                    _encodedPath[i] = socketAddress[s_nativePathOffset + i];
+                }
+
+                // Strip trailing null of pathname socket addresses.
+                int length = _encodedPath.Length;
+                if (!IsAbstract(_encodedPath))
+                {
+                    // Since this isn't an abstract path, we're sure our first byte isn't 0.
+                    while (_encodedPath[length - 1] == 0)
+                    {
+                        length--;
+                    }
+                }
+                _path = s_pathEncoding.GetString(_encodedPath, 0, length);
+            }
+            else
+            {
+                _encodedPath = Array.Empty<byte>();
+                _path = string.Empty;
+            }
+        }
+
+        public override SocketAddress Serialize()
+        {
+            var result = new SocketAddress(AddressFamily.Unix, s_nativeAddressSize);
```

Comment:
The length changed from `s_nativePathOffset + _encodedPath.Length` to `s_nativeAddressSize`.

Reply:
Why the change?

---
In reply to: [171722363](https://github.com/dotnet/corefx/pull/27631#discussion_r171722363) [](ancestors = 171722363)

Reply:
Because it needs to match the native type in size/layout; otherwise if the native implementation dereferences, it can AV and other such things, and the native implementation may also do a size check. Basically it's dangerous to use anything smaller than the corresponding native definition.